### PR TITLE
Production-level support for MSSQL

### PR DESCRIPTION
**Pull request by potiuk**
MSSQL has been somewhat experimental in the `main` branch, but as
we near releasing for 2.2.0 version, the image should support
the mssql at the level as it supports other databases.

This PR adds proper support for both PROD and CI images.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by mik-laj** [comment]
Could MSSQL be used as a celery backend?
https://github.com/apache/airflow/blob/f76eaec65e834f9c8c3e4e0bec14bb63e76679b8/scripts/in_container/prod/entrypoint_prod.sh#L98-L106

**Comment by potiuk** [comment]
> Could MSSQL be used as a celery backend?

Very good point! Indeed I believe it can

**Comment by potiuk** [comment]
> Could MSSQL be used as a celery backend?

Added support for the entrypoint . Good catch @mik-laj !

File: Dockerfile.ci [hunk 1]
```diff
@@ -104,6 +104,7 @@ RUN mkdir -pv /usr/share/man/man1 \
 
 COPY scripts/docker/*.sh /scripts/docker/
 RUN bash /scripts/docker/install_mysql.sh dev \
+    && /scripts/docker/install_mssql.sh \
```
Comment (mik-laj) [inline Dockerfile.ci]:
```suggestion
    && bash /scripts/docker/install_mssql.sh \
```
For consistency. See previous line.

File: scripts/docker/install_mssql.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+set -euo pipefail
+
+set -x
```
Comment (mik-laj) [inline scripts/docker/install_mssql.sh]: Why this script has so many ``set`` statements?

**Comment by mik-laj** [comment]
I am missing one more element. On [the database configuration](https://github.com/apache/airflow/blob/main/docs/apache-airflow/howto/set-up-database.rst) page, we should add an example URL address.

For SQLite, we have the following fragment:

 > An example URI for the sqlite database:
> 
> sqlite:////home/airflow/airflow.db

For MySQL, we have the following fragment:

> We recommend using the mysqlclient driver and specifying it in your SqlAlchemy connection string.
> 
> mysql+mysqldb://<user>:<password>@<host>[:<port>]/<dbname>
> But we also support the mysql-connector-python driver, which lets you connect through SSL without any cert options provided.
> 
> mysql+mysqlconnector://<user>:<password>@<host>[:<port>]/<dbname>

For. PostgresSQL, we have the following fragment:

> We recommend using the psycopg2 driver and specifying it in your SqlAlchemy connection string.
> 
> postgresql+psycopg2://<user>:<password>@<host>/<db>

Only MSSQL does not include an example URI address for me to be able to. configure the database. SQLAlchemy supports multiple drivers. Which one is the best for production deployment?

```
# pyodbc
mssql+pyodbc://scott:tiger@mydsn

# pymssql
mssql+pymssql://scott:tiger@hostname:port/dbname

# mxODBC
mssql+mxodbc://<username>:<password>@<dsnname>
```

File: Dockerfile.ci (hunk 1 above)
Comment (potiuk) [reply Dockerfile.ci]: Right!

File: scripts/docker/install_mssql.sh (hunk 2 above)
Comment (potiuk) [reply scripts/docker/install_mssql.sh]: Good point - no idea. Same in mysql - remnant of an old refactor which was missed I guessed.

**Comment by potiuk** [comment]
> Only MSSQL does not include an example URI address for me to be able to. configure the database. SQLAlchemy supports multiple drivers. Which one is the best for production deployment?

@aneesh-joseph  ?

**Comment by potiuk** [comment]
> Only MSSQL does not include an example URI address for me to be able to. configure the database. SQLAlchemy supports multiple drivers. Which one is the best for production deployment?

I added mssql+pyodbc as this is the one a) we support in the the image b) we test during CI. Since this is rather fresh - I will leave it at that until we have an evidence that other drivers should be preferred (but then we should also change our tests and image to use them).

**Comment by potiuk** [comment]
good comments - All solved @mik-laj

**Comment by mik-laj** [comment]
@potiuk SQLAlchemy published some recommendations about preferred drivers:  
PyODBC has the best Unicode support.
https://docs.sqlalchemy.org/en/14/dialects/mssql.html#driver-unicode-support

> Deprecated since version 1.4: The mxODBC DBAPI is deprecated and will be removed in a future version. Please use one of the supported DBAPIs to connect to mssql.

https://docs.sqlalchemy.org/en/14/dialects/mssql.html#dialect-mssql-mxodbc-connect

>pymssql is currently not included in SQLAlchemy’s continuous integration (CI) testing.

https://docs.sqlalchemy.org/en/14/dialects/mssql.html#module-sqlalchemy.dialects.mssql.pymssql

**Review by mik-laj — APPROVED** [review]

File: docs/apache-airflow/howto/set-up-database.rst [hunk 3]
```diff
@@ -248,6 +248,13 @@ You can read more about transaction isolation and snapshot features at
    GRANT ALL PRIVILEGES ON DATABASE airflow TO airflow_user;
 
 
+We recommend using the ``mssql+pyodbc`` driver and specifying it in your SqlAlchemy connection string.
+
+.. code-block:: text
+
+    mssql+pyodbc://<user>:<password>@<host>
```
Comment (mik-laj) [inline docs/apache-airflow/howto/set-up-database.rst:255]: On CI, we use a little more complex URI. This leads me to suppose that this example may be incomplete. https://github.com/apache/airflow/blob/c686241f4ceb62d52e9bfa607822e4b7a3c76222/scripts/ci/docker-compose/backend-mssql.yml#L23-L24

Comment (potiuk) [reply docs/apache-airflow/howto/set-up-database.rst:255]: This is the default one from the sqlalchemy docs (I took it from there):

https://docs.sqlalchemy.org/en/14/core/engines.html#microsoft-sql-server 

We use mor complex one, but those are optional parameters, and I wouldn't dive into details of it (especially not before we have first users using 'for real' and making comments about it.

Comment (mik-laj) [reply docs/apache-airflow/howto/set-up-database.rst:255]: Should we add a database name to the example? I think it is a commonly used parameter. It needed to set up a database created in the paragraph above. 

```
mssql+pyodbc://<user>:<password>@<host>
mssql+pyodbc://<user>:<password>@<host>/<database_name>
```

Comment (mik-laj) [reply docs/apache-airflow/howto/set-up-database.rst:255]: I created a PR: https://github.com/apache/airflow/pull/18404

**Comment by ashb** [comment]
How much does this add to the docker images size btw?

**Comment by potiuk** [comment]
> How much does this add to the docker images size btw?

It increases the size of the image by 4MB - from 972 to 976. See the discussion here: https://apache-airflow.slack.com/archives/CQAMHKWSJ/p1632131354042400

**Pull request merged**